src/test/java/org/mastodon/mamut/model/ModelUndoRedoTest.java

```diff
@@ -0,0 +1,276 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+package org.mastodon.mamut.model;
+
+import static org.junit.Assert.assertEquals;
+
+import java.awt.Color;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.StringJoiner;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.junit.Test;
+import org.mastodon.mamut.io.importer.ModelImporter;
+import org.mastodon.model.tag.TagSetStructure;
+import org.mastodon.util.TagSetUtils;
+
+/**
+ * The robustness of the undo and redo methods is crucial for Mastodon. This
+ * test thoroughly tests the undo and redo methods of the {@link Model} class.
+ */
+public class ModelUndoRedoTest
+{
+	/**
+	 * Test if undo and redo works for basic operations: adding spots and edges.
+	 */
+	@Test
+	public void testUndoRedo()
+	{
+		Model model = new Model();
+		ModelGraph graph = model.getGraph();
+		model.setUndoPoint();
+		Spot a = graph.addVertex().init( 0, new double[ 3 ], 1 );
+		a.setLabel( "A" );
+		model.setUndoPoint();
+		assertEquals( "A", graphAsString( graph ) );
+		Spot b = graph.addVertex().init( 0, new double[ 3 ], 1 );
+		b.setLabel( "B" );
+		model.setUndoPoint();
+		assertEquals( "A, B", graphAsString( graph ) );
+		graph.addEdge( a, b ).init();
+		model.setUndoPoint();
+		assertEquals( "A, B, A->B", graphAsString( graph ) );
+		model.undo();
+		assertEquals( "A, B", graphAsString( graph ) );
+		model.undo();
+		assertEquals( "A", graphAsString( graph ) );
+		model.undo();
+		assertEquals( "", graphAsString( graph ) );
+		model.redo();
+		assertEquals( "A", graphAsString( graph ) );
+		model.redo();
+		assertEquals( "A, B", graphAsString( graph ) );
+		model.redo();
+		assertEquals( "A, B, A->B", graphAsString( graph ) );
+	}
+
+	/**
+	 * Returns a readable sting representation of the graph.
+	 */
+	private String graphAsString( ModelGraph graph )
+	{
+		List< String > spots = new ArrayList<>();
+		for ( Spot spot : graph.vertices() )
+			spots.add( spot.getLabel() );
+		Collections.sort( spots );
+		List< String > links = new ArrayList<>();
+		for ( Link link : graph.edges() )
+			links.add( link.getSource().getLabel() + "->" + link.getTarget().getLabel() );
+		Collections.sort( links );
+		StringJoiner joiner = new StringJoiner( ", " );
+		spots.forEach( joiner::add );
+		links.forEach( joiner::add );
+		return joiner.toString();
+	}
+
+	/**
+	 * This test makes various changes to the model and then tests if the undo
+	 * and redo methods work as expected.
+	 */
+	@Test
+	public void testUndoRedoOnEmptyModel()
+	{
+		Model model = new Model();
+		testUndoRedo( model );
+	}
+
+	/**
+	 * This method makes various changes to the graph and the tag set model. The
+	 * changes are divided into several steps, and at each step, an undo point is
+	 * set. Then, the undo and redo methods are called to test if the model is
+	 * restored to the intended state.
+	 */
+	private static void testUndoRedo( Model model )
```

Comment:
```suggestion
	private static void testUndoRedo( final Model model )
```
It seems like you forgot the final modifiers for the variables. It is not strictly speaking required, but seems to be a convention in this project.